src/app/reporting/Engine.cpp

```diff
@@ -104,28 +99,44 @@ CHIP_ERROR Engine::BuildSingleReportDataAttributeDataList(ReportData::Builder &
     aReportDataBuilder.Checkpoint(backup);
     AttributeDataList::Builder attributeDataList = aReportDataBuilder.CreateAttributeDataListBuilder();
     SuccessOrExit(err = aReportDataBuilder.GetError());
-    // TODO: Need to handle multiple chunk of message
-    for (auto clusterInfo = apReadHandler->GetAttributeClusterInfolist(); clusterInfo != nullptr; clusterInfo = clusterInfo->mpNext)
+
+    if (apReadHandler->IsInitialReport())
     {
-        if (apReadHandler->IsInitialReport())
+        ConcreteAttributePath path;
+        mMoreChunkedMessages = true;
+        for (; apReadHandler->GetPathIterator()->Get(path); apReadHandler->GetPathIterator()->Proceed())
         {
+            TLV::TLVWriter attributeBackup;
+            aReportDataBuilder.Checkpoint(attributeBackup);
             // Retrieve data for this cluster instance and clear its dirty flag.
-            err = RetrieveClusterData(apReadHandler->GetFabricIndex(), attributeDataList, *clusterInfo);
+            err = RetrieveClusterData(apReadHandler->GetFabricIndex(), attributeDataList, path);
+            if (err != CHIP_NO_ERROR)
+            {
+                aReportDataBuilder.Rollback(attributeBackup);
+            }
             VerifyOrExit(err == CHIP_NO_ERROR,
                          ChipLogError(DataManagement, "<RE:Run> Error retrieving data from cluster, aborting"));
             attributeClean = false;
         }
-        else
+        mMoreChunkedMessages = false;
+    }
+    else
+    {
+        for (auto clusterInfo = apReadHandler->GetAttributeClusterInfolist(); clusterInfo != nullptr;
+             clusterInfo      = clusterInfo->mpNext)
         {
             for (auto path = mpGlobalDirtySet; path != nullptr; path = path->mpNext)
             {
-                if (clusterInfo->IsAttributePathSupersetOf(*path))
+                if (clusterInfo->IsAttributePathSupersetOf(*path) && !path->HasWildcard())
```

Comment:
This logic doesn't look quite right - what if one of the dirty paths was a wildcard? In that case, this wouldn't handle it right?

e.g Read EP*, and the dirty path contains EP1/ClusterA/AttributeId*. In this case, this check would fail, and so would the next else if statement.

The correct approach would be to take the resultant intersection (as you're doing with the two if else branches here), THEN initialize a path iterator on that intersection and iterate through that like you're doing above.